_posts/2021-08-02-Musing-on-Microservices.md

```diff
@@ -0,0 +1,29 @@
+---
+layout: post
+title: Musing on Microservices
+author: Charles Thomas
+---
+
+Microservices are commonplace among technology companies of all sizes and in this post, I hope to explore some of the pros and cons of them and what you should consider when thinking about whether or not they are a good solution for the problems you're solving.
+
+Let's start by quickly addressing a common question: what are microservices? Microservices are a way of structuring your application so that instead of having one big application (referred to as a monolith), you have multiple smaller applications that communicate with each other over a network.
+
+So why might you want to do this? Generally, microservices set out to solve two problems: scalability (how can I make my application handle more load?) and availability (how can I keep my application serving requests more of the time?). Let's deal with each of these issues separately.
```

Comment:
There are definitely other problems they set out to solve:

- Individual deployability (isolating scope of bad changes, decreasing deployment surface area, easier rollbacks etc.)
- Abstraction (expose behaviour via a common protocol, developers are free to use any language / framework etc. as they please)

Probably a load of others I'm skipping

Reply:
I would argue that individual deployability isn't really something that comes with microservices. It comes with having good build tools and deploying frequently. You can deploy small changes and roll back quickly with most monoliths. If anything rolling back is much harder with microservices and you have to ensure compatibility between all service versions